ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Background.hs

```diff
@@ -0,0 +1,344 @@
+-- concurrently. To enforce thread-safety, a lock is obtained at the start of
+-- this function and released at the end. So in practice, this function can be
+-- called multiple times concurrently, but the calls will be serialised.
+--
+-- NOTE: this function /can/ run concurrently with all other functions, just
+-- not with itself.
+copyToImmDB
+  :: forall m blk.
+     ( MonadSTM   m
+     , MonadCatch m
+     , OuroborosTag (BlockProtocol blk)
+     , HasHeader blk
+     , HasHeader (Header blk)
+     )
+  => ChainDbEnv m blk
+  -> m SlotNo
+copyToImmDB CDB{..} = withCopyLock $ do
+    toCopy <- atomically $ do
+      curChain <- readTVar cdbChain
+      let nbToCopy = max 0 (AF.length curChain - fromIntegral k)
+          toCopy :: [Point blk]
+          toCopy = map headerPoint
+                 $ take nbToCopy
+                 $ AF.toOldestFirst curChain
+      return toCopy
+
+    if null toCopy
+      then trace NoBlocksToCopyToImmDB
+      else forM_ toCopy $ \pt -> do
+        let hash = case pointHash pt of
+              BlockHash h -> h
+              -- There is no actual genesis block that can occur on a chain
+              GenesisHash -> error "genesis block on current chain"
+        -- This call is cheap
+        slotNoAtImmDBTip <- ImmDB.getSlotNoAtTip cdbImmDB
+        if pointSlot pt > slotNoAtImmDBTip
+          then do
+            blk <- VolDB.getKnownBlock cdbVolDB hash
+            -- We're the only one modifying the ImmutableDB, so the tip cannot
+            -- have changed since we last checked it.
+            ImmDB.appendBlock cdbImmDB blk
+            -- TODO the invariant of 'cdbChain' is shortly violated between
+            -- these two lines: the tip was updated on the line above, but the
+            -- anchor point is only updated on the line below.
+            atomically $ removeFromChain pt
+            trace $ CopiedBlockToImmDB pt
+          else error "block to copy was older than the ImmutableDB's tip"
+
+    -- Get the /possibly/ updated tip of the ImmDB
+    ImmDB.getSlotNoAtTip cdbImmDB
+  where
+    SecurityParam k = protocolSecurityParam cdbNodeConfig
+    trace = traceWith (contramap TraceCopyToImmDBEvent cdbTracer)
+
+    -- | Remove the header corresponding to the given point from the beginning
+    -- of the current chain fragment.
+    --
+    -- PRECONDITION: the header must be the first one (oldest) in the chain
+    removeFromChain :: Point blk -> STM m ()
+    removeFromChain pt = do
+      -- The chain might have been extended in the meantime.
+      curChain <- readTVar cdbChain
+      case curChain of
+        hdr :< curChain'
+          | headerPoint hdr == pt
+          -> writeTVar cdbChain curChain'
+        -- We're the only one removing things from 'curChain', so this cannot
+        -- happen if the precondition was satisfied.
+        _ -> error "header to remove not on the current chain"
+
+    withCopyLock :: forall a. m a -> m a
```

Comment:
If we don't expect this to ever run concurrently with itself, throw an exception/error if it happens.